source/agora/node/Validator.d

```diff
@@ -411,4 +416,50 @@ public class Validator : FullNode, API
         super.shutdown();
         this.nominator.storeLatestState();
     }
+
+    /***************************************************************************
+
+        Check if the current enrollment is about to expire or validator is not
+        enrolled. Enroll when necessary.
+
+        Params:
+            block_height = Current block height
+
+    ***************************************************************************/
+
+    private void checkAndEnroll (Height block_height) @safe
+    {
+        Hash enroll_key = this.enroll_man.getEnrolledUTXO(this.utxo_set.getUTXOFinder());
+
+        if (enroll_key == Hash.init && (enroll_key = this.getFrozenUTXO()) == Hash.init)
+            return; // Not enrolled and no frozen UTXO
+
+        const enrolled = this.enroll_man.getEnrolledHeight(enroll_key);
+
+        // This validators enrollment will expire next cycle or not enrolled at all
```

Comment:
How can `enrolled` be `ulong.max` here? Shouldn't the previous `if` block take care of that?

Reply:
It would be `ulong.max` when the Validator is not enrolled but has a Frozen UTXO.

Reply:
But the comment above says `Not enrolled and no frozen UTXO`

Reply:
Yes, so the early return handles the case where validator is not currently enrolled and there is no frozen utxo that we can use to enroll. If we dont return early we definitely have a frozen utxo. We will send an enrollment request in 2 cases, if the validator is not enrolled or the current enrollment is about to expire. `enrolled == ulong.max` covers the first case, `block_height + 1 >= enrolled + this.params.ValidatorCycle` covers the second.

Reply:
Ah yes it makes sense now. Am a bit slow today. :)

Reply:
I am wondering if `checkAndEnroll()` should be called in `start()` or somewhere. I am trying to see if that would be a problem if validator havent caught up with the network yet.

Reply:
Thanks.  You made me notice a problem with our enrollments: https://github.com/bpfkorea/agora/issues/1297 :)

Reply:
Accidental contribution!